### PR TITLE
support oceanbase

### PR DESCRIPTION
**Pull request by bashaojing**

### CODE REVIEW

**Comment by oraluben** [comment]
Do you intend to add a full test workflow for OceanBase? (install OceanBase, run test, etc.)

File: .travis.yml [hunk 1]
```diff
@@ -144,3 +144,9 @@ matrix:
         - sleep 5
       script:
         - CLICKHOUSE_AVAILABLE=true mvn -Dtest=ClickHouseBinaryComparisonOperationTest test
+    - name: OceanBase
```
Comment (mrigger) [inline .travis.yml:147]: We are now actually using GitHub Actions for testing. For example, see https://github.com/sqlancer/sqlancer/blob/master/.github/workflows/main.yml#L187 for how the MySQL tests are executed in the CI. If you want, you can also add this in a follow-up PR.

File: src/sqlancer/oceanbase/OceanBaseErrors.java [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+        errors.add("Incorrect decimal value");
+        errors.add("error parsing regexp");
+        errors.add("The observer or zone is not the master");
+        errors.add("Incorrect integer value");
+        errors.add("Truncated incorrect DOUBLE value");
+        errors.add("Data truncated for argument");
+        errors.add("Invalid numeric");
+
+
+        if (true) {
```
Comment (mrigger) [inline src/sqlancer/oceanbase/OceanBaseErrors.java:44]: I guess we can remove the `if` and only keep its body?

File: src/sqlancer/oceanbase/README.md [hunk 3]
```diff
@@ -0,0 +1,21 @@
+## Install Oceanbase
```
Comment (mrigger) [inline src/sqlancer/oceanbase/README.md:1]: That's very useful! Thanks for adding also a `README.md`.

File: src/sqlancer/oceanbase/ast/OceanBaseConstant.java [hunk 4]
```diff
@@ -0,0 +1,628 @@
...
+            }    
+        }
+
+        @Override
+        public boolean asBooleanNotNull() {
+            for (int i = value.length(); i >= 1; i--) {
+                try {
+                    char currentChar = value.charAt(i-1);
+                    int currentVal= Integer.valueOf(currentChar);
+                    if (currentVal < 48||currentVal > 57)
```
Comment (mrigger) [inline src/sqlancer/oceanbase/ast/OceanBaseConstant.java:214]: Having named constants for 48 and 57 would be useful (but I assume that one of the style checks will complain about this anyway).

File: src/sqlancer/oceanbase/oracle/OceanBaseNoRECOracle.java [hunk 5]
```diff
@@ -0,0 +1,204 @@
...
+
+        OceanBaseUnaryPostfixOperation isNotFalse= new OceanBaseUnaryPostfixOperation(randomWhereCondition,OceanBaseUnaryPostfixOperation.UnaryPostfixOperator.IS_FALSE,true);
+
+        OceanBaseUnaryPostfixOperation isNULL= new OceanBaseUnaryPostfixOperation(randomWhereCondition,OceanBaseUnaryPostfixOperation.UnaryPostfixOperator.IS_NULL,false);
+
+        OceanBaseUnaryPostfixOperation isNotNULL= new OceanBaseUnaryPostfixOperation(randomWhereCondition,OceanBaseUnaryPostfixOperation.UnaryPostfixOperator.IS_NULL,true);
+
+        OceanBaseExpression expr = OceanBaseConstant.createNullConstant(); 
+        switch(i){
+           case 0:
```
Comment (mrigger) [inline src/sqlancer/oceanbase/oracle/OceanBaseNoRECOracle.java:165]: The hard-coded constants look quite error-prone. Perhaps you can add an enum instead, of which a random value is selected, and based on which you perform the switch (or where each enum value implements a method that you call)?

File: src/sqlancer/oceanbase/oracle/OceanBaseNoRECOracle.java [hunk 6]
```diff
@@ -0,0 +1,204 @@
...
+            return -1;
+        }
+        if (rs.next()) {
+            firstCount += rs.getLong(1);
+        }
+        rs.close();
+        return firstCount;
+    }
+  
+    private OceanBaseExpression getTrueExpr(OceanBaseExpression randomWhereCondition){
```
Comment (mrigger) [inline src/sqlancer/oceanbase/oracle/OceanBaseNoRECOracle.java:149]: Perhaps you can add a short comment explaining how this differs from the standard NoREC version?

**Review by mrigger — COMMENTED** [review]
Thanks a lot for the PR! I think we can merge it soon. I added some minor comments. I would appreciate it if you would address them before we merge the PR. Currently, also some style checks are failing. You can check locally using `mvn verify`.

**Review by mrigger — APPROVED** [review]
LGTM! Thanks a lot again for the PR! It would be ideal to contribute a GitHub Actions workflow to test the OceanBase implementation to ensure that no future changes break it. Do you plan to also contribute such a workflow?

**Comment by bashaojing** [comment]
> 

> LGTM! Thanks a lot again for the PR! It would be ideal to contribute a GitHub Actions workflow to test the OceanBase implementation to ensure that no future changes break it. Do you plan to also contribute such a workflow?

Thanks for you approval! It needs some preceding conditions  for installing oceanbase. I'm not sure installing oceanbase will always success when running ci and block other  commit.

**Comment by mrigger** [comment]
If the installation would work on the current version, I think it would be good enough. We regularly adapt to updates to database systems. For example, H2 currently fails due to an update, which I hope to address soon.

**Comment by mrigger** [comment]
Another issue: I noticed that building SQLancer now results in multiple warnings:
``` 
sqlancer/src/sqlancer/oceanbase/ast/OceanBaseConstant.java:[96,24] The constructor java.lang.Double(double) is deprecated
[WARNING] sqlancer/src/sqlancer/oceanbase/ast/OceanBaseConstant.java:[107,34] The constructor java.lang.Double(double) is deprecated
[WARNING] sqlancer/src/sqlancer/oceanbase/ast/OceanBaseConstant.java:[113,34] The constructor java.lang.Double(double) is deprecated
[WARNING] sqlancer/src/sqlancer/oceanbase/ast/OceanBaseConstant.java:[281,54] Dead code
[WARNING] sqlancer/src/sqlancer/oceanbase/ast/OceanBaseConstant.java:[305,50] Dead code
[WARNING] sqlancer/src/sqlancer/oceanbase/gen/datadef/OceanBaseIndexGenerator.java:[85,18] The type sqlancer.oceanbase.gen.datadef.OceanBaseIndexGenerator.PartitionOptions is never used locally
[INFO] sqlancer/src/sqlancer/oceanbase/ast/OceanBaseComputableFunction.java:[33,38] Varargs methods should only override or be overridden by other varargs methods unlike new sqlancer.oceanbase.ast.OceanBaseComputableFunction.OceanBaseFunction(){}.apply(sqlancer.oceanbase.ast.OceanBaseConstant[], sqlancer.oceanbase.ast.OceanBaseExpression[]) and sqlancer.oceanbase.ast.OceanBaseComputableFunction.OceanBaseFunction.apply(sqlancer.oceanbase.ast.OceanBaseConstant[], sqlancer.oceanbase.ast.OceanBaseExpression...)
[INFO] sqlancer/src/sqlancer/oceanbase/ast/OceanBaseComputableFunction.java:[47,38] Varargs methods should only override or be overridden by other varargs methods unlike new sqlancer.oceanbase.ast.OceanBaseComputableFunction.OceanBaseFunction(){}.apply(sqlancer.oceanbase.ast.OceanBaseConstant[], sqlancer.oceanbase.ast.OceanBaseExpression[]) and sqlancer.oceanbase.ast.OceanBaseComputableFunction.OceanBaseFunction.apply(sqlancer.oceanbase.ast.OceanBaseConstant[], sqlancer.oceanbase.ast.OceanBaseExpression...)
[INFO] sqlancer/src/sqlancer/oceanbase/ast/OceanBaseComputableFunction.java:[67,38] Varargs methods should only override or be overridden by other varargs methods unlike new sqlancer.oceanbase.ast.OceanBaseComputableFunction.OceanBaseFunction(){}.apply(sqlancer.oceanbase.ast.OceanBaseConstant[], sqlancer.oceanbase.ast.OceanBaseExpression[]) and sqlancer.oceanbase.ast.OceanBaseComputableFunction.OceanBaseFunction.apply(sqlancer.oceanbase.ast.OceanBaseConstant[], sqlancer.oceanbase.ast.OceanBaseExpression...)
[INFO] sqlancer/src/sqlancer/oceanbase/ast/OceanBaseComputableFunction.java:[84,38] Varargs methods should only override or be overridden by other varargs methods unlike new sqlancer.oceanbase.ast.OceanBaseComputableFunction.OceanBaseFunction(){}.apply(sqlancer.oceanbase.ast.OceanBaseConstant[], sqlancer.oceanbase.ast.OceanBaseExpression[]) and sqlancer.oceanbase.ast.OceanBaseComputableFunction.OceanBaseFunction.apply(sqlancer.oceanbase.ast.OceanBaseConstant[], sqlancer.oceanbase.ast.OceanBaseExpression...)
[INFO] sqlancer/src/sqlancer/oceanbase/ast/OceanBaseComputableFunction.java:[98,38] Varargs methods should only override or be overridden by other varargs methods unlike new sqlancer.oceanbase.ast.OceanBaseComputableFunction.OceanBaseFunction(){}.apply(sqlancer.oceanbase.ast.OceanBaseConstant[], sqlancer.oceanbase.ast.OceanBaseExpression[]) and sqlancer.oceanbase.ast.OceanBaseComputableFunction.OceanBaseFunction.apply(sqlancer.oceanbase.ast.OceanBaseConstant[], sqlancer.oceanbase.ast.OceanBaseExpression...)
[INFO] sqlancer/src/sqlancer/oceanbase/ast/OceanBaseComputableFunction.java:[105,38] Varargs methods should only override or be overridden by other varargs methods unlike new sqlancer.oceanbase.ast.OceanBaseComputableFunction.OceanBaseFunction(){}.apply(sqlancer.oceanbase.ast.OceanBaseConstant[], sqlancer.oceanbase.ast.OceanBaseExpression[]) and sqlancer.oceanbase.ast.OceanBaseComputableFunction.OceanBaseFunction.apply(sqlancer.oceanbase.ast.OceanBaseConstant[], sqlancer.oceanbase.ast.OceanBaseExpression...)
[WARNING] sqlancer/src/sqlancer/oceanbase/OceanBaseUserCheckException.java:[3,14] The serializable class OceanBaseUserCheckException does not declare a static final serialVersionUID field of type long
``` 
Could you please fix them?

**Pull request merged**